### PR TITLE
Add ZeroBaseFeeMarket to support post-London free gas networks

### PR DESCRIPTION
**Pull request by siladu**
Details on https://github.com/hyperledger/besu/issues/4061

This allows private networks already using London fee market to start using free gas. Previously, since https://github.com/hyperledger/besu/pull/4003, it was only possible to start a free gas London network from genesis.

It also fixes an issue where the 0 initial base fee can creep back up to non-zero when a block uses enough gas: https://github.com/hyperledger/besu/issues/4218

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Review by antonydenyer — APPROVED** [review]

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -1408,6 +1408,10 @@ public void run() {
         if (genesisConfigOptions.isQuorum()) {
           enableGoQuorumCompatibilityMode();
         }
+
+        if (genesisConfigOptions.isZeroBaseFee()) {
```
Comment (shemnon) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1412]: Is this truly needed?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java [hunk 2]
```diff
@@ -39,6 +40,11 @@ public BaseFeeMarketBlockHeaderGasPriceValidationRule(final BaseFeeMarket baseFe
   public boolean validate(final BlockHeader header, final BlockHeader parent) {
     try {
 
+      if (baseFeeMarket instanceof ZeroBaseFeeMarket) {
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java:43]: Design wise this is too tight of coupling.  Instead there should be a method in `BaseFeeMarket` `isValidatesBasefee` that returns true or false based on a market config.

Or `isForkBlock` should be removed and replaced with a `validationMode(long blocknumber)` method that returns an enum with 3 states - `NONE`, `INITIAL`, and `ONGOING`.  Zero market would return `NONE`

**Review by shemnon — CHANGES_REQUESTED** [review]
Design wise there is too tight a coupling with an instanceof check for a specific market instance - https://github.com/hyperledger/besu/pull/4174/files#diff-69d6b308e912a669614fc09ad1224590d582389dfc56698f94cd63a4ad68e33eR43.  Change that to a solution that doesn't require an instanceof check.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 1 above)
Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1412]: Not really, it's more of a convenience. I'm happy to remove.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java (hunk 2 above)
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java]: I like the validationMode idea, went with that 👍

**Review by shemnon — APPROVED** [review]

**Comment by siladu** [comment]
Converting to a draft to delay merging this in until after The Merge release

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java [hunk 3]
```diff
@@ -40,16 +40,22 @@ public boolean validate(final BlockHeader header, final BlockHeader parent) {
...
-      if (baseFeeMarket.isForkBlock(header.getNumber())) {
+      if (BaseFeeMarket.ValidationMode.INITIAL.equals(
+          baseFeeMarket.baseFeeValidationMode(header.getNumber()))) {
         return baseFeeMarket
             .getInitialBasefee()
             .equals(header.getBaseFee().orElseThrow(() -> MissingBaseFeeFromBlockHeader()));
+      } else if (BaseFeeMarket.ValidationMode.NONE.equals(
+          baseFeeMarket.baseFeeValidationMode(header.getNumber()))) {
+        // skip validation to allow sync when a zero baseFee chain has historical baseFees
+        return true;
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java:51]: if ZeroBaseFeeMarket overrode getInitialBaseFee and always returned zero, we could leave this conditional as it was?

Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/headervalidationrules/BaseFeeMarketBlockHeaderGasPriceValidationRule.java:51]: The scenario solved by this conditional is when you have some baseFee present on your network's block (enabled by London) and then subsequently set `zeroBaseFee: true`. So your blocks' baseFee goes from empty -> non-zero -> 0

During syncing, this rule ends up trying to compare the expected base fee (zero, due to overridden computeBaseFee) with the actual one (non-zero)

**Review by garyschulte — APPROVED** [review]
LGTM, 22.7.3-next 👍

**Pull request merged**